4_basic_assembly/2_bitwise_operations/2_practical_bit_games/answers/3_write_code/3.asm

```diff
@@ -0,0 +1,87 @@
+; Basic Assembly
+; ==============
+;
+; 3.  Sum of distances:
+;    
+;     For a binary number x, we define the "sum of 1 distances" to be the sum of
+;     distances between every two "1" bits in x's binary representation.
+;
+;     Small example (8 bits):
+; 
+;       x = {10010100}_2
+;            7  4 2
+;      
+;       The total sum of distances: (7-4) + (7-2) + (4-2) = 3 + 5 + 2 = 10
+;
+;     Create a program that takes a number x (of size 4 bytes) as input, and
+;     outputs x's "sum of 1 distances".
+;
+; Based on SO: 
+; https://stackoverflow.com/questions/32470046/calculating-sum-of-all-the-distance/32470309#32470309
+
+format PE console
+entry start
+
+include 'win32a.inc' 
+
+; ===============================================
+section '.text' code readable executable
+
+start:
+
+    call    read_hex
+    mov     ecx,32d
+    mov     edx,0
+
+; count the number of ones in a number
+sum_bits:
+    mov     ebx,eax
+    and     ebx,1
+    add     edx,ebx
+    ror     eax,1
+    loop    sum_bits
+
+; calculate the sum of all 1 distances
+    mov     ebx,edx
+    mov     ecx,32d
+    mov     edx,0
+    mov     esi,0
+    mov     edi,0

```

Comment:
It can be very helpful if we add here comments for edx, esi, edi and ebx. This comment should be something like:
```asm
mov ebx, edx     ; ebx is remaining 1 bits to check.
xor edx, edx     ; the previous position of 1
...
```
More than helping the reader, this should be of great use to the "future you" when you read this code again.